Code Entry Types documentation (#1239)
nuclio/nuclio: docs/reference/function-configuration/code-entry-types.md

# Code Entry Types Reference

This document describes the various function code entry types.

#### In This Document

- [Basic Code Entry](#basic-code-entry)
- [External Code Entries](#external-code-entries)
- [Image Code Entries](#image-code-entry)
- [See also](#see-also)

## Basic code entry

> In the basic code entry, the function configuration and the source code are taken from the passed spec. here's an example:

Reviewer: ```suggestion
In basic code entry type, the function configuration and the source code are taken directly from the spec. The source code is encoded directly in `build.functionSourceCode` field as a base64 string. For example:
```

> (The source code is passed as base64 string)

Reviewer: ```suggestion
```

```yaml
spec:
  description: my Go function
  handler: main:Handler
  runtime: golang
  build:
    functionSourceCode: "cGFja2FnZSBtYWluDQoNCmltcG9ydCAoDQogICAgImdpdGh1Yi5jb20vbnVjbGlvL251Y2xpby1zZGstZ28iDQopDQoNCmZ1bmMgSGFuZGxlcihjb250ZXh0ICpudWNsaW8uQ29udGV4dCwgZXZlbnQgbnVjbGlvLkV2ZW50KSAoaW50ZXJmYWNle30sIGVycm9yKSB7DQogICAgcmV0dXJuIG5pbCwgbmlsDQp9"
    codeEntryType: "sourceCode"
```

> ## External code entries

Reviewer: ```suggestion
## External code entry types
```

There are two types of external code entries - Github and Archive.
When external code entry type is set, the function source code and configuration are being fetched from the code entry.
When deploying this way, the source code files will be fetched from the code entry and the function configuration will 
get enriched with the code entry's function configuration.
* The code entry's function configuration values will not override the passed configuration values.
* List values will be merged (such as meta.annotations, meta.labels etc...).
* The environment variables will also be merged.

> * When using external code entry all of the source code files are saved and can be used by the handler.

Reviewer: ```
Today,  two external code entry types are supported - Github and Archive.
When deploying a function with an external code entry type, the source code files will be fetched from the remote target - via github api or fetching and uncompressing the archive from the given url.
The source code will be taken from the given external source, and the given function configuration from the request will be enriched with the function configuration appearing in the remote code location. The configuration merge will give precedence to the function configuration passed over the remote function configuration fields, using a merge strategy.
* Passed configuration values will override configuration in the remote source code (archive or other)
* List values will be merged (such as meta.annotations, meta.labels etc...), In particular, environment variables will be merged.
* When using external code entry all of the source code files are saved and can be used by the handler.
* External code entry types make use of 2 `codeEntryAttributes` - a `headers` text2text map and a `workDir` string
```
